adrs/2023-02-10-limit-manager-role-permission.md

```diff
@@ -0,0 +1,79 @@
+
+- The cache-based API client requires cluster scope `list` and `watch` permission for any resource the controller may query.
+
+- This documentation only scopes to the AutoscalingRunnerSet CRD and its controller.
+
+## Service accounts and their role binding in actions-runner-controller
+
+There are 3 service accounts involved for a working `AutoscalingRunnerSet` based `actions-runner-controller`
+
+1. Service account for each Ephemeral runner Pod
+
+This should have the lowest privilege (not any `RoleBinding` nor `ClusterRoleBinding`) by default, in the case of `containerMode=kubernetes`, it will get certain write permission with `RoleBinding` to limit the permission to a single namespace.
+
+> References:
+> - ./charts/auto-scaling-runner-set/templates/no_permission_serviceaccount.yaml
+> - ./charts/auto-scaling-runner-set/templates/kube_mode_role.yaml
+> - ./charts/auto-scaling-runner-set/templates/kube_mode_role_binding.yaml
+> - ./charts/auto-scaling-runner-set/templates/kube_mode_serviceaccount.yaml
+
+2. Service account for AutoScalingListener Pod
+
+This has a `RoleBinding` to a single namespace with a role that has permission to `PATCH` `EphemeralRunnerSet` and `EphemeralRunner`.
+
+3. Service account for the controller manager
+
+Since the CRD controller is a singleton installed in the cluster that manages the CRD across multiple namespaces, the service account of the controller manager pod has a `ClusterRoleBinding` to a `ClusterRole` with broader permissions.
+
+The current `ClusterRole` has the following permissions:
+
+- Get/List/Create/Delete/Update/Patch/Watch on `AutoScalingRunnerSets` (with `Status` and `Finalizer` sub-resource)
+- Get/List/Create/Delete/Update/Patch/Watch on `AutoScalingListeners` (with `Status` and `Finalizer` sub-resource)
+- Get/List/Create/Delete/Update/Patch/Watch on `EphemeralRunnerSets` (with `Status` and `Finalizer` sub-resource)
+- Get/List/Create/Delete/Update/Patch/Watch on `EphemeralRunners` (with `Status` and `Finalizer` sub-resource)
+
+- Get/List/Create/Delete/Update/Patch/Watch on `Pods` (with `Status` sub-resource)
+- **Get/List/Create/Delete/Update/Patch/Watch on `Secrets`**
+- Get/List/Create/Delete/Update/Patch/Watch on `Roles`
+- Get/List/Create/Delete/Update/Patch/Watch on `RoleBindings`
+- Get/List/Create/Delete/Update/Patch/Watch on `ServiceAccounts`
+
+> Full list can be found at: https://github.com/actions/actions-runner-controller/blob/facae69e0b189d3b5dd659f36df8a829516d2896/charts/actions-runner-controller-2/templates/manager_role.yaml
+
+## Limit cluster role permission on Secrets
+
+The cluster scope `List` `Secrets` permission might be a blocker for adopting `actions-runner-controller` for certain customers as they may have certain restriction in their cluster that simply doesn't allow any service account to have cluster scope `List Secrets` permission. 
```

Comment:
This is just "my two cents" right now but anyway- another possibility that isn't covered in the current proposal looks to be about introducing a "single namespace" mode to the new ScaleSet-based ARC and using RoleBinding instead of ClusterRoleBinding.

That would allow us to still leverage the list and watch APIs for performance.

> Note that controller-runtime has the non-cached and the cached client. The former does not depend on the list and watch APIs while the latter scales better.

Also, note that the legacy ARC had `singleNamespace=true` and `watchNamespace=NS_NAME` chart values to enable the single namespace mode. The legacy chart is [hard-coded to use ClusterRoleBinding anyway
(https://github.com/actions/actions-runner-controller/blob/master/charts/actions-runner-controller/templates/manager_role_binding.yaml) so we might better use RoleBinding instead for the single namespace mode.

Perhaps the same approach would work for the new ARC too?

Reply:
I am trying to see if we eliminate the need for List secrets in cluster scope, how much do we need to support the watch single namespace installation?

Reply:
do you have an idea of the scenarios that the current users try to install ARC with `singleNamespace=true`?

Reply:
It was so people could deploy multiple controllers allowing them to work around rate limit restrictions by using multiple PATs / GitHub Apps. It preceded the introduction of caching so possibly isn't as required for large installations as it once was when first introduced but we wouldn't know.

Reply:
Thanks for the context @toast-gear, I would avoid introducing the single namespace installation for the new `AutoScalingRunnerSet` mode for now since the new mode suppose to handle a much higher load without consuming all your GitHub API rate limit. 🙇 
We can always revisit this base on the customer's ask. 😄

Reply:
Thanks for sharing @toast-gear! That matches my memory too.

@TingluoHuang To be clear, although the single namespace mode has been added for scalability, the idea here is it could be leveraged to enhance the security posture of the legacy and new ARC by two means-

First, it can be configured to use the namespaced controller-runtime client(which is backed by the namespaced List and Watch API calls).

Second, we can use role bindings instead of cluster role bindings for tighter access control.

This way, you have layered security- the potential attacker needs to break two things instead of one or zero in a relevant attack scenario before they can grab random secrets contained in the cluster.

Code auditing (https://github.com/actions/actions-runner-controller/pull/2275#discussion_r1106255247 ) is important too so that you can see "how" secure ARC is. On the other hand, I believe layered security is about making ARC "more" secure.

We can probably continue the discussion about this "what API verbs on which namespaces to support in ARC" thing in @wherka-ama's thread below? https://github.com/actions/actions-runner-controller/pull/2275#discussion_r1106160657

Reply:
> Thanks for sharing @toast-gear! That matches my memory too.
> 
> @TingluoHuang To be clear, although the single namespace mode has been added for scalability, the idea here is it could be leveraged to enhance the security posture of the legacy and new ARC by two means-
> 
> First, it can be configured to use the namespaced controller-runtime client(which is backed by the namespaced List and Watch API calls).
> 
> Second, we can use role bindings instead of cluster role bindings for tighter access control.
> 
> This way, you have layered security- the potential attacker needs to break two things instead of one or zero in a relevant attack scenario before they can grab random secrets contained in the cluster.
> 
> Code auditing ([#2275 (comment)](https://github.com/actions/actions-runner-controller/pull/2275#discussion_r1106255247) ) is important too so that you can see "how" secure ARC is. On the other hand, I believe layered security is about making ARC "more" secure.
> 
> We can probably continue the discussion about this "what API verbs on which namespaces to support in ARC" thing in @wherka-ama's thread below? [#2275 (comment)](https://github.com/actions/actions-runner-controller/pull/2275#discussion_r1106160657)

@mumoshu : I couldn't agree more with your statement here:
"This way, you have layered security- the potential attacker needs to break two things instead of one or zero in a relevant attack scenario before they can grab random secrets contained in the cluster."

Thanks a lot for recognising the need to work on that subject. We really appreciate that!